amr-wind/fvm/nonLinearSum.H

```diff
@@ -0,0 +1,161 @@
+#ifndef NONLINEARSUM_H
+#define NONLINEARSUM_H
+
+#include "amr-wind/fvm/fvm_utils.H"
+
+namespace amr_wind::fvm {
+
+/** Non-linear Term for Kosovic Model
+ *  \ingroup fvm
+ */
+template <typename FTypeIn, typename FTypeOut>
+struct NonLinearSum
+{
+    NonLinearSum(FTypeOut& strphi, const FTypeIn& phi)
+        : m_strphi(strphi), m_phi(phi)
+    {
+        AMREX_ALWAYS_ASSERT(AMREX_SPACEDIM == m_phi.num_comp());
+    }
+
+    template <typename Stencil>
+    void apply(const int lev, const amrex::MFIter& mfi) const
+    {
+        const auto& geom = m_phi.repo().mesh().Geom(lev);
+        const auto& idx = geom.InvCellSizeArray();
+        const auto& strphi = m_strphi(lev).array(mfi);
+        const auto& phi = m_phi(lev).const_array(mfi);
+
+        const auto& bx_in = mfi.tilebox();
+        const auto& bx = Stencil::box(bx_in, geom);
+        if (bx.isEmpty()) {
+            return;
+        }
+
+        amrex::ParallelFor(
+            bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                amrex::Real cp1, c, cm1, ux, uy, uz, vx, vy, vz, wx, wy, wz;
+                cp1 = Stencil::c00;
+                c = Stencil::c01;
+                cm1 = Stencil::c02;
+
+                ux = (cp1 * phi(i + 1, j, k, 0) + c * phi(i, j, k, 0) +
+                      cm1 * phi(i - 1, j, k, 0)) *
+                     idx[0];
+                vx = (cp1 * phi(i + 1, j, k, 1) + c * phi(i, j, k, 1) +
+                      cm1 * phi(i - 1, j, k, 1)) *
+                     idx[0];
+                wx = (cp1 * phi(i + 1, j, k, 2) + c * phi(i, j, k, 2) +
+                      cm1 * phi(i - 1, j, k, 2)) *
+                     idx[0];
+
+                cp1 = Stencil::c10;
+                c = Stencil::c11;
+                cm1 = Stencil::c12;
+
+                uy = (cp1 * phi(i, j + 1, k, 0) + c * phi(i, j, k, 0) +
+                      cm1 * phi(i, j - 1, k, 0)) *
+                     idx[1];
+                vy = (cp1 * phi(i, j + 1, k, 1) + c * phi(i, j, k, 1) +
+                      cm1 * phi(i, j - 1, k, 1)) *
+                     idx[1];
+                wy = (cp1 * phi(i, j + 1, k, 2) + c * phi(i, j, k, 2) +
+                      cm1 * phi(i, j - 1, k, 2)) *
+                     idx[1];
+
+                cp1 = Stencil::c20;
+                c = Stencil::c21;
+                cm1 = Stencil::c22;
+
+                uz = (cp1 * phi(i, j, k + 1, 0) + c * phi(i, j, k, 0) +
+                      cm1 * phi(i, j, k - 1, 0)) *
+                     idx[2];
+                vz = (cp1 * phi(i, j, k + 1, 1) + c * phi(i, j, k, 1) +
+                      cm1 * phi(i, j, k - 1, 1)) *
+                     idx[2];
+                wz = (cp1 * phi(i, j, k + 1, 2) + c * phi(i, j, k, 2) +
+                      cm1 * phi(i, j, k - 1, 2)) *
+                     idx[2];
+
+                // N11
+                strphi(i, j, k, 0) = (ux * ux + uy * vx + uz * wx) -
```

Comment:
I know this is a lot to ask, but for readability (and for fewer bugs), I am wondering if this might look better when left as a tensor expression instead of hand unrolling it. 

Something along the lines of:
```
    for (int ii = 0; ii < AMREX_SPACEDIM; ++ii) {
        for (int jj = 0; jj < AMREX_SPACEDIM; ++jj) {
            for (int kk = 0; kk < AMREX_SPACEDIM; ++kk) {
                N(i,j,k,ii*AMREX_SPACEDIM + jj) += grad_phi(i,j,k,ii,kk) * grad_phi(i,j,k,kk,ii)...
```

I'm still on the fence about this, since I'm not sure what the performance difference between loops and hand unrolled expressions will be (my gut feeling with modern compilers is that it shouldn't matter).

Reply:
I unrolled them to ensure that the paper written terms got coded up correctly into the code.

Reply:
This is purely stylistic and I'm not necessarily asking you to change the code for the PR; but I think it's clearer if you have the tensor expressions written directly, see below as an example https://github.com/Exawind/amr-wind/blob/0bf6748359c51fa1beb2b74832ed09d680cd0285/amr-wind/turbulence/LES/AMDNoTherm.H#L51-L80.